dotnet/src/InternalUtilities/connectors/Memory/MongoDB/MongoDBGenericDataModelMapper.cs

```diff
@@ -38,27 +26,30 @@ public BsonDocument MapFromDataToStorageModel(VectorStoreGenericDataModel<string
         var document = new BsonDocument();
 
         // Loop through all known properties and map each from the data model to the storage model.
-        foreach (var property in this._vectorStoreRecordDefinition.Properties)
+        foreach (var property in model.Properties)
         {
-            var storagePropertyName = property.StoragePropertyName ?? property.DataModelPropertyName;
-
-            if (property is VectorStoreRecordKeyProperty keyProperty)
+            switch (property)
             {
-                document[MongoDBConstants.MongoReservedKeyPropertyName] = dataModel.Key;
-            }
-            else if (property is VectorStoreRecordDataProperty dataProperty)
-            {
-                if (dataModel.Data is not null && dataModel.Data.TryGetValue(dataProperty.DataModelPropertyName, out var dataValue))
-                {
-                    document[storagePropertyName] = BsonValue.Create(dataValue);
-                }
-            }
-            else if (property is VectorStoreRecordVectorProperty vectorProperty)
-            {
-                if (dataModel.Vectors is not null && dataModel.Vectors.TryGetValue(vectorProperty.DataModelPropertyName, out var vectorValue))
-                {
-                    document[storagePropertyName] = BsonArray.Create(GetVectorArray(vectorValue));
-                }
+                case VectorStoreRecordKeyPropertyModel keyProperty:
+                    document[MongoDBConstants.MongoReservedKeyPropertyName] = dataModel.Key;
+                    continue;
+
+                case VectorStoreRecordDataPropertyModel dataProperty:
+                    if (dataModel.Data is not null && dataModel.Data.TryGetValue(dataProperty.ModelName, out var dataValue))
+                    {
+                        document[dataProperty.StorageName] = BsonValue.Create(dataValue);
+                    }
+                    continue;
+
+                case VectorStoreRecordVectorPropertyModel vectorProperty:
+                    if (dataModel.Vectors is not null && dataModel.Vectors.TryGetValue(vectorProperty.ModelName, out var vectorValue))
+                    {
+                        document[property.StorageName] = BsonArray.Create(GetVectorArray(vectorValue));
+                    }
+                    continue;
+
+                default:
+                    throw new UnreachableException();
```

Comment:
just an observation: since for certain targets we include `UnreachableException` as  link, the users won't be able to catch it with `catch (UnreachableException ex)` because they are going to have their own copy defined in a different library.

FWIW I would use `InvalidDataException` here